clang/lib/AST/ExprConstant.cpp

```diff
@@ -13595,6 +13595,15 @@ bool IntExprEvaluator::VisitUnaryExprOrTypeTraitExpr(
                     Info.Ctx.getOpenMPDefaultSimdAlign(E->getArgumentType()))
             .getQuantity(),
         E);
+  case UETT_VectorElements: {
+    QualType Ty = E->getTypeOfArgument();
+    // If the vector has a fixed size, we can determine the number of elements
+    // at compile time.
+    if (Ty->isVectorType())
+      return Success(Ty->castAs<VectorType>()->getNumElements(), E);
+
+    return false;
```

Comment:
Shouldn't this 'return false' just be an  assert?  (That is, instead of the 'if')?  Or is there a way for a non-vector type to make it through?

Additionally, what about sizeless vectors here?  Should we be testing for those instead to make this a 'fail'?  And does that case need a diagnostic?

Reply:
maybe my understanding here is wrong, so please correct me in that case. If the vector is fixed-sized, we can determine the number of elements at compile time, which is why we return the number here in `ExprConstant.cpp`. if it is a scalable vector, we `return false`, as we cannot determine a constant value. in that case, we move along and hit `CGExprScalar.cpp`, which is where we emit the runtime `vscale` code for anything that is not constant. 

So in my understanding, we either emit a constant in `ExprConstant` or the runtime logic in `CGExprScalar`. I can add assertions to check that we are always in a good case, but i've not come across a fixed-sized vector path that does not get caught in `ExprConstant`. if there are alternative code paths, then you are correct and I should add checks. but I don't want to duplicate logic if this case can never happen anyway.

Reply:
Your understanding of how this works is correct, I missed you were handling sizeless vectors that way.  So above 'return false' there 
could be an `assert(Ty->isSizelessVectorType())` ?  

Note that we're not guaranteed to come to ExprConstant.cpp in the case where it is a sized-vector, so you likely need to make sure the CGExprScalar code handles sized types as well.